collections/src/main/scala/strawman/collection/Iterator.scala

```diff
@@ -1133,31 +1133,51 @@ trait Iterator[+A] extends IterableOnce[A] { self =>
 
 }
 
-object Iterator {
+object Iterator extends IterableFactory[Iterator] {
 
-  val empty: Iterator[Nothing] = new Iterator[Nothing] {
+  private[this] val _empty: Iterator[Nothing] = new Iterator[Nothing] {
     def hasNext = false
     def next() = throw new NoSuchElementException("next on empty iterator")
   }
 
+  /** Creates a target $coll from an existing source collection
+    *
+    * @param source Source collection
+    * @tparam A the type of the collection’s elements
+    * @return a new $coll with the elements of `source`
+    */
+  override def from[A](source: IterableOnce[A]): Iterator[A] = source.iterator()
+
+  /** The iterator which produces no values. */
+  @`inline` def empty[T]: Iterator[T] = _empty
```

Comment:
inlining this method has little effect
If the method is inlinable it requires the caller to be monomorphically dispatched, which would imply that you have you pretty redundant code in this case

what you need to do in inline the target access. See https://github.com/rorygraves/scalac_perf/issues/46
maybe `@`inline` def empty[T]: Iterator[T] = _empty: @inline `
but you will need to check the bytecode :-(
@szeiger can you consider where else this should be applied, or discuss with @retronym if the optimiser should do this automatically

Reply:
All implementations below:

```scala
@inline final def empty[T]: Iterator[T] = _empty: @inline
@inline final def empty[T]: Iterator[T] = _empty
def empty[T]: Iterator[T] = _empty
```

result in the following bytecode:

```java
  public static <T> strawman.collection.Iterator<T> empty();
    Code:
       0: getstatic     #66                 // Field strawman/collection/Iterator$.MODULE$:Lstrawman/collection/Iterator$;
       3: invokevirtual #146                // Method strawman/collection/Iterator$.empty:()Lstrawman/collection/Iterator;
       6: areturn
```

As per client code is concerned I tried definition

```scala
@inline final def empty[T]: Iterator[T] = _empty: @inline
```

and the following client code:

```scala
object IteratorInlineTest {

  def inlineTest1: immutable.List[scala.Int] = immutable.List.from(Iterator.empty)

  def inlineTest2(iter: Iterator[scala.Int] = Iterator.empty): Iterator[scala.Int] =
    Iterator(iter, Iterator(3, 21), Iterator(1, 2, 34))
      .foldLeft(Iterator.empty[scala.Int])(_ ++ _)
  
  def inlineTest3: Iterator[scala.Int] = {
    var itr = Iterator.empty[Int]
    itr = itr ++ Iterator(123)
    itr = itr ++ Iterator(321)
    itr
  }
}
```

which resulted with the following bytecode:

```java
public final class strawman.collection.IteratorInlineTest {
  public static strawman.collection.Iterator<java.lang.Object> inlineTest2$default$1();
    Code:
       0: getstatic     #16                 // Field strawman/collection/IteratorInlineTest$.MODULE$:Lstrawman/collection/IteratorInlineTest$;
       3: invokevirtual #18                 // Method strawman/collection/IteratorInlineTest$.inlineTest2$default$1:()Lstrawman/collection/Iterator;
       6: areturn

  public static strawman.collection.Iterator<java.lang.Object> inlineTest3();
    Code:
       0: getstatic     #16                 // Field strawman/collection/IteratorInlineTest$.MODULE$:Lstrawman/collection/IteratorInlineTest$;
       3: invokevirtual #21                 // Method strawman/collection/IteratorInlineTest$.inlineTest3:()Lstrawman/collection/Iterator;
       6: areturn

  public static strawman.collection.Iterator<java.lang.Object> inlineTest2(strawman.collection.Iterator<java.lang.Object>);
    Code:
       0: getstatic     #16                 // Field strawman/collection/IteratorInlineTest$.MODULE$:Lstrawman/collection/IteratorInlineTest$;
       3: aload_0
       4: invokevirtual #25                 // Method strawman/collection/IteratorInlineTest$.inlineTest2:(Lstrawman/collection/Iterator;)Lstrawman/collection/Iterator;
       7: areturn

  public static strawman.collection.immutable.List<java.lang.Object> inlineTest1();
    Code:
       0: getstatic     #16                 // Field strawman/collection/IteratorInlineTest$.MODULE$:Lstrawman/collection/IteratorInlineTest$;
       3: invokevirtual #29                 // Method strawman/collection/IteratorInlineTest$.inlineTest1:()Lstrawman/collection/immutable/List;
       6: areturn
}
```

and actually I get the same bytecode in the case of current definition. Please advice if any other case has to be taken into consideration apart from these ones above.